### PR TITLE
8320370: NMT: Change MallocMemorySnapshot to simplify code.

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi,

`MallocMemorySnapshot` used to be initialized in a quite non-standard way, using global placement new on an array of `size_t` of the size required to fit `MallocMemorySnapshot`. This looks like it was intended to circumvent some `ResourceObj` internals, but I'm unsure of its purpose. This change does what you expect a regular initialization of a global variable to look like.

~Currently running through tier1.~ GHA passed and so did tier1 on Oracle CI.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320370](https://bugs.openjdk.org/browse/JDK-8320370): NMT: Change MallocMemorySnapshot to simplify code. (**Enhancement** - P4)

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [b0c547b4](https://git.openjdk.org/jdk/pull/16724/files/b0c547b444fe3b5dcbbbd6e0a0c17fd0c417dde9)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16724/head:pull/16724` \
`$ git checkout pull/16724`

Update a local copy of the PR: \
`$ git checkout pull/16724` \
`$ git pull https://git.openjdk.org/jdk.git pull/16724/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16724`

View PR using the GUI difftool: \
`$ git pr show -t 16724`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16724.diff">https://git.openjdk.org/jdk/pull/16724.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16724#issuecomment-1818614476)

### CODE REVIEW

File: src/hotspot/share/nmt/mallocTracker.cpp [hunk 1]
```diff
@@ -43,7 +43,7 @@
 #include "utilities/ostream.hpp"
 #include "utilities/vmError.hpp"
 
-size_t MallocMemorySummary::_snapshot[CALC_OBJ_SIZE_IN_TYPE(MallocMemorySnapshot, size_t)];
+MallocMemorySnapshot MallocMemorySummary::_snapshot{};
```
Comment (tstuefe) [inline src/hotspot/share/nmt/mallocTracker.cpp:46]: Why the curly braces?

File: src/hotspot/share/nmt/virtualMemoryTracker.cpp [hunk 2]
```diff
@@ -49,9 +49,6 @@ void VirtualMemory::update_peak(size_t size) {
 #endif // ASSERT
 
 void VirtualMemorySummary::initialize() {
-  assert(sizeof(_snapshot) >= sizeof(VirtualMemorySnapshot), "Sanity Check");
-  // Use placement operator new to initialize static data area.
-  ::new ((void*)_snapshot) VirtualMemorySnapshot();
 }
```
Comment (tstuefe) [inline src/hotspot/share/nmt/virtualMemoryTracker.cpp:52]: Can be removed.

**Review by tstuefe — COMMENTED** [review]
Looks good, smaller nits.

File: src/hotspot/share/nmt/virtualMemoryTracker.cpp (hunk 2 above)
Comment (jdksjolen) [reply src/hotspot/share/nmt/virtualMemoryTracker.cpp]: Oh, thanks! Should've seen that :-).

File: src/hotspot/share/nmt/mallocTracker.cpp (hunk 1 above)
Comment (jdksjolen) [reply src/hotspot/share/nmt/mallocTracker.cpp:46]: They're not necessary, I just have a tendency to explicitly use brace/"uniform" initialization everywhere. I've removed them.

**Review by tstuefe — APPROVED** [review]
Good

**Comment by tstuefe** [comment]
I believe the reason for the original shape of this code was to avoid initialization altogether if NMT was disabled. @zhengyu123 ?

**Review by gerard-ziemski — APPROVED** [review]
Nice clean up!

Just one question: I'm probably missing something here, but did we have to change from `ResourceObj` to `AnyObj` for this change to work?

**Comment by jdksjolen** [comment]
> Nice clean up!
> 
> Just one question: I'm probably missing something here, but did we have to change from `ResourceObj` to `AnyObj` for this change to work?

Hi Gerard, no I think it is unnecessary to do this change to make it work (I haven't checked). I do think keeping it as `AnyObj` is a good idea however, as `MallocMemorySnapshot` is never resource allocated.

**Comment by jdksjolen** [comment]
> > > Nice clean up!
> > > Just one question: I'm probably missing something here, but did we have to change from `ResourceObj` to `AnyObj` for this change to work?
> > 
> > 
> > Hi Gerard, no I think it is unnecessary to do this change to make it work (I haven't checked). I do think keeping it as `AnyObj` is a good idea however, as `MallocMemorySnapshot` is never resource allocated.
> 
> Why not make it a simple object then? Why derive from anything at all?
> 
> (Update: I am fine with the current state of the patch; up to you)

It's not clear to me what the rules are on this matter. To what degree are plain global classes allowed in HotSpot? I thought that they all need to inherit from `AnyObj`,`CHeapObj`, etc. Seems like we're more OK with internal (structs and classes within other classes) being "undecorated" with their allocation capabilities.

**Comment by tstuefe** [comment]
> > > > Nice clean up!
> > > > Just one question: I'm probably missing something here, but did we have to change from `ResourceObj` to `AnyObj` for this change to work?
> > > 
> > > 
> > > Hi Gerard, no I think it is unnecessary to do this change to make it work (I haven't checked). I do think keeping it as `AnyObj` is a good idea however, as `MallocMemorySnapshot` is never resource allocated.
> > 
> > 
> > Why not make it a simple object then? Why derive from anything at all?
> > (Update: I am fine with the current state of the patch; up to you)
> 
> It's not clear to me what the rules are on this matter. To what degree are plain global classes allowed in HotSpot? I thought that they all need to inherit from `AnyObj`,`CHeapObj`, etc. Seems like we're more OK with internal (structs and classes within other classes) being "undecorated" with their allocation capabilities.

I would hope so. Why increase compiler load and force inclusion of unrelated headers for stuff that is really not necessary.

**Comment by zhengyu123** [comment]
> I believe the reason for the original shape of this code was to avoid initialization altogether if NMT was disabled. @zhengyu123 ?

The original code was to workaround 

> I believe the reason for the original shape of this code was to avoid initialization altogether if NMT was disabled. @zhengyu123 ?

I believe it was coded this way to avoid memory allocation (malloc) during NMT initialization, before tracking infrastructure is ready, e.g. `MallocSiteTable` is initialized.

**Comment by stefank** [comment]
> > > Nice clean up!
> > > Just one question: I'm probably missing something here, but did we have to change from `ResourceObj` to `AnyObj` for this change to work?
> > 
> > 
> > Hi Gerard, no I think it is unnecessary to do this change to make it work (I haven't checked). I do think keeping it as `AnyObj` is a good idea however, as `MallocMemorySnapshot` is never resource allocated.
> 
> Why not make it a simple object then? Why derive from anything at all?
> 
> (Update: I am fine with the current state of the patch; up to you)

+1 Please don't use AnyObj unless you really need multiple allocation strategies.

**Comment by tstuefe** [comment]
> > I believe the reason for the original shape of this code was to avoid initialization altogether if NMT was disabled. @zhengyu123 ?
> 
> I believe it was coded this way to avoid memory allocation (malloc) during NMT initialization, before tracking infrastructure is ready, e.g. `MallocSiteTable` is initialized.
> 
> IIRC, not using static variable was to workaround the order of static variable initialization by linker, cannot remember the details.

That makes a lot of sense. Its a moot point now since https://bugs.openjdk.org/browse/JDK-8256844

**Review by tstuefe — APPROVED** [review]
Still looks good. Nice simplificaiton.

**Comment by jdksjolen** [comment]
> Still looks good. Nice simplificaiton.

Thanks for the re-approval!

> > > > Nice clean up!
> > > > Just one question: I'm probably missing something here, but did we have to change from `ResourceObj` to `AnyObj` for this change to work?
> > > 
> > > 
> > > Hi Gerard, no I think it is unnecessary to do this change to make it work (I haven't checked). I do think keeping it as `AnyObj` is a good idea however, as `MallocMemorySnapshot` is never resource allocated.
> > 
> > 
> > Why not make it a simple object then? Why derive from anything at all?
> > (Update: I am fine with the current state of the patch; up to you)
> 
> +1 Please don't use AnyObj unless you really need multiple allocation strategies.

I'm very happy to hear this :).

@gerard-ziemski, would you mind re-approving also?

**Comment by gerard-ziemski** [comment]
> > Still looks good. Nice simplificaiton.
> 
> Thanks for the re-approval!
> 
> > > > > Nice clean up!
> > > > > Just one question: I'm probably missing something here, but did we have to change from `ResourceObj` to `AnyObj` for this change to work?
> > > > 
> > > > 
> > > > Hi Gerard, no I think it is unnecessary to do this change to make it work (I haven't checked). I do think keeping it as `AnyObj` is a good idea however, as `MallocMemorySnapshot` is never resource allocated.
> > > 
> > > 
> > > Why not make it a simple object then? Why derive from anything at all?
> > > (Update: I am fine with the current state of the patch; up to you)
> > 
> > 
> > +1 Please don't use AnyObj unless you really need multiple allocation strategies.
> 
> I'm very happy to hear this :).
> 
> @gerard-ziemski, would you mind re-approving also?

We currently have:

`class MallocMemorySnapshot : public AnyObj {`

which used to be:

`class MallocMemorySnapshot : public ResourceObj {`

Are we OK with using `AnyObj` here?

**Comment by tstuefe** [comment]
I'm fine with it either way, but would prefer the object not to inherit from anything if its only used inline or as stack object. In any case, lets ship this.

**Comment by jdksjolen** [comment]
Hi,

Keeping the `as_snapshot` but removed the inheritance. Thanks for your patience with this PR.

**Review by stefank — APPROVED** [review]

**Review by gerard-ziemski — APPROVED** [review]

**Pull request closed without merging**